<p style="text-align: center;"><font size="8"><b>Functions and Modules</b></font><br>


# Calling Functions

Pure functions exist as methods that can be called outside the context of a particular class. For example we're already seen the `round` and `len` functions. Remember that we call `round(a)`, not `a.round()`. Python provides many built-in functions. 

![functions](https://github.com/lukasbystricky/ISC-3313/blob/master/lectures/chapter2/images/functions.png?raw=true)

In [1]:
pow(2,3)

8

In [2]:
min(-2,-2.4,9.0)

-2.4

In [3]:
ord("a")

97

# Print function

After performing a computation we need to see the result. In an interactive or notebook session we can simply type the name of the variable to see what it is.

In [4]:
a = 1
a

1

When executing scripts however, we can no longer to this. Instead we use the `print` command (in fact we've already used this several times).

The `print` function is used to print information to the console. 

In [5]:
print(a)
print("a")

1
a


Note the syntax used in the textbook, `print a`, (no parentheses) does not work in Python 3. However the syntax above works in both Python 2 and Python 3.


Of course, `print` can be used to display more useful information. For example, suppose we have a variable called `t` that represents some length of time in seconds. We can use `print` to display not only `t`, but also the units by calling `print` with multiple arguments.

In [6]:
t = 5.6
print(t, "s")

5.6 s


The `print` function automatically inserts a space between the two arguments. If we wish the avoid this, we can combine the two arguments into 1.

In [7]:
print(str(t)+"s")

5.6s


Note that we have to convert `round(t,1)` to a string first before "adding" (the correct term is *concatenating*) it to `s`. The command

In [8]:
print(t+"s")

TypeError: unsupported operand type(s) for +: 'float' and 'str'

is illegal because we are attempting to add a `float` to a `str`.

# Modules

All the functions listed above are built-in to Python. This means they are automatically avaliable once we start Python. There are hundreds of other useful functions and classes that have been developped for Python which are not automatically loaded, but instead placed into specialized libraries called *modules* that can be individually loaded as needed.

## Math module

The math module provides functions to do mathematical operations beyond addition, multiplication, exponentiation etc. For example suppose we want to take the square root of a number. This is not built-in the Python, however a function called `sqrt` in the math library does this. 

To use the `sqrt` function we must import it from the math module. There are three possible ways to do this. Whichever method you use should be done at the beginning of your code (this is not strictly speaking necessary, but it is the most common placement, at the very least you have to do this *before* you use the function).

1) Import the entire module. 

In [10]:
import math

At this point we still cannot use the `sqrt` command directly, we must specifically tell Python where the function is coming from using a *qualified name*.

In [11]:
sqrt(2)

NameError: name 'sqrt' is not defined

In [12]:
math.sqrt(2)

1.4142135623730951

2) Specifically importing `sqrt` from the math library. If we are using `sqrt` several times in the code, this can avoid repeated typing.

In [13]:
from math import sqrt
sqrt(2)

1.4142135623730951

3) Import everything in the module, avoid using qualified names for all functions. This can be an attractive option but is generally discouraged because different modules may use the same name for different functions. This method of importing imports not only the `sqrt` function, but all functions from the math library, for example `cos` and `sin`.

In [14]:
from math import *
sqrt(2)

1.4142135623730951

In [15]:
sin(pi/2)

1.0

# Expressions

Before looking at other modules, let's look quickly at expressions. We have already seen several expressions in isolation (e.g. 18+5.5).

It is quite common the perform several operations as part of a single expression.

In [16]:
a = 18 + 5.5 + 1
a

24.5

In this case behind the scenes Python adds 18 and 5.5 to get 23.5, it then adds 1 to get 24.5. In this case the order of the two operations does not matter, however in more complicated expressions the order can be important.

In [17]:
a = 18*9**2/4
a

364.5

In [18]:
a = 9/4**2*18
a

10.125

## Precedence

When there are two or more operations as part of an expression, we must figure out some way to determine which operation is performed first. We say that an operation that is performed fist is given *precedence* over the others.

Mathematical expressions in Python follow standard algebraic conventions:
1. Brackets
2. Exponents
3. Division/Multiplication
4. Addition/Subtraction

For example in the expression `1 + 2*3` the multiplication is done first, followed by the addition. 

In Python, as in algebra we can use brackets to prioritize an operation.

In [19]:
1+2*3

7

In [20]:
(1+2)*3

9

Most operations with equal precedence are evaluated left to right, again to mimic standard algebraic rules.  

One exception is exponenents which are evaluated right to left, which again is how we typically think of exponents.

$4^{3^2} = 4^9$

In [21]:
4**3**2

262144

In [22]:
4**(3**2)

262144

In [23]:
(4**3)**2

4096

Even though precedence rules are based on algebraic rules, they are enforced for any data type. 

In [24]:
"a"*3+"b"

'aaab'

### Example

When boiling an egg, it has been determined that the time it takes for the center of the yolk to reach a desired temperature $T$ is given by
$$ t = \frac{M^{2/3}c\rho^{1/3}}{K\pi^2(4\pi/3)^{2/3}}\ln\left(0.76\frac{T_0 - 100}{T - 100}\right)$$

where
* $M$ is the mass of the egg
* $\rho$ is the density
* $c$ is the specific heat capacity
* $K$ is the thermal conductivity
* $T_0$ is the temperature at $t=0$

$$ t = \frac{M^{2/3}c\rho^{1/3}}{K\pi^2(4\pi/3)^{2/3}}\ln\left(0.76\frac{T_0 - 100}{T - 100}\right)$$

In [25]:
T = 70 # desired temperature
M = 47
rho = 1.038
c = 3.7
K = 5.4e-3
T0 = 4

# compute time according to above formula, we need the natural logarithm function from the math module
from math import log

t = (M**(2/3)*c*rho**(1/3))/(K*pi**2*(4*pi/3)**(2/3))*log(0.76*(T0 - 100)/(T - 100))
t

313.09454902221637

## Calling Functions from Within Expressions

Function calls have high precedence. When multiple finction calls are used in the same expression they are typically evaluated from left to right.

In [26]:
person = "George Washignton"
person.split()[1]

'Washignton'

More complicated expressions are evaluated by first resolving commands inside parentheses.

In [27]:
groceries = ["cereal", "milk", "apple"]
groceries.insert(groceries.index("milk") + 1, "eggs")
groceries

['cereal', 'milk', 'eggs', 'apple']

Here we first must evaluate `groceries.index("milk")` and then add 1 to it to find the index where we wish to insert "eggs".

# NumPy

The `NumPy` module (http://www.numpy.org/) is an almost indespensible module for scientific computing. It provides objects such as arrays and matrices as well as functions spanning linear algebra, fourier transforms and stastics amoung numerous other things. 

Almost every assignment in this course will use NumPy. Here we will present a very basic application that demonstrates some of the NumPy function, namely evaluating the integral of a function.

To start with we must import NumPy. To reduce the amount of typing for ourselves later we will call the module `np`.

In [30]:
import numpy as np

## Arrays

One important class that NumPy provides is the `array` class. An array is similar to a `list` in that it is a collection of objects. Typically arrays store numbers. 

NumPy arrays can be initialized in a similar way to lists.

In [31]:
a = np.array([1,2.0,3])
type(a)

numpy.ndarray

You'll notice the type of a is `numpy.ndarray`. NumPy arrays can be multidimensional. Mathematically we can think of array of dimension 1 as a vector, and an array of dimension 2 as a matrix. Higher dimensional arrays are certainly possible. The attribute `ndim` stores the number of dimensions in the array.

In [32]:
a.ndim

1

Multidimensional arrays can be initialized as follows: 

In [33]:
b = np.array([[1, 2, 3.0], [1.2,2.2,2]])
b.ndim

2

It should be noted that 2D arrays are __not__ matrices. For example, if we call:

In [34]:
x = np.array([1,2])
A = np.array([[3,2], [1,2]])

print(A*x)

[[3 4]
 [1 4]]


We get a 2D array, instead of a vector. This is because array operations are defined elementwise. $A\mathbf{x}$ in this case is defined to be:

$$ \begin{bmatrix} A_{11}x_1 & A_{12}x_2\\ A_{21}x_1 & A_{22} x_2\end{bmatrix}$$

Likewise if we call $A^2$, we get

In [35]:
print(A**2)

[[9 4]
 [1 4]]


which is 
$$ \begin{bmatrix} A_{11}^2 & A_{12}^2\\A_{21}^2 & A_{22}^2\end{bmatrix}.$$
This is not what we expect from matrix squaring.

Fortunately NumPy has a matrix class that acts like we'd expect a matrix to act.

In [45]:
B = np.matrix([[3,2],[1,2]])
x = np.array([1,2])
B**2

matrix([[11, 10],
        [ 5,  6]])

To do matrix vector (or vector vector) products, we must use the `np.dot` command:

In [50]:
print(B.dot(x))

[[7 5]]


Like lists and strings, arrays support indexing and slicing.

In [ ]:
a[0]

In [ ]:
b[1]

In [ ]:
b[1][0:2]

### Arrays vs. Lists

Arrays and lists are similar in many ways. Both represent a collection of objects. When should you use one over the other? 

For starters, arrays are mutable, however they do not support methods such as `pop` or `append`. Once initialized the size of an array cannot be easily changed. If your application needs to change the size of a collection, lists are the prefered option. 

In general lists are prefered unless there is a specific reason to use NumPy arrays. NumPy arrays are less flexible; however you'll find that they are needed to interact fully with the rest of NumPy. For example taking the dot product of two vectors requires the two vectors to be stored as NumPy arrays. 

It's possible to convert from a list to an array or vice versa. NumPy arrays provide the method `tolist()` which converts an array to a list.

In [ ]:
c = a.tolist()
type(c)

NumPy also provides the function `asarray` that takes a list and returns an array.

In [ ]:
d = np.asarray(c)
type(d)